src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -434,11 +435,34 @@ private void SafeSendResponse(Message message, object resultObject)
             }
         }
 
+        private static readonly Lazy<Func<Exception, Exception>> prepForRemotingLazy = 
+            new Lazy<Func<Exception, Exception>>(() =>
+            {
+                ParameterExpression exceptionParameter = Expression.Parameter(typeof(Exception));
+                MethodCallExpression prepForRemotingCall = Expression.Call(exceptionParameter, "PrepForRemoting", Type.EmptyTypes);
+                Expression<Func<Exception, Exception>> lambda = Expression.Lambda<Func<Exception, Exception>>(prepForRemotingCall, exceptionParameter);
+                Func<Exception, Exception> func = lambda.Compile();
+                return func;
+            });
+
+        private static Exception PrepareForRemoting(Exception exception)
+        {
+            // Call the Exception.PrepForRemoting internal method, which preserves the original stack when the exception
+            // is rethrown at the remote site (and appends the call site stacktrace). If this is not done, then when the
+            // exception is rethrown the original stacktrace is entire replaced.
+            // Note: another commonly used approach since .NET 4.5 is to use ExceptionDispatchInfo.Capture(ex).Throw()
+            // but that involves rethrowing the exception in-place, which is not what we want here, but could in theory
```

Comment:
Because this is the remote site before we send it through the wire. We need to rethrow the exception at the other end of the wire. Doable, but requires some extra non trivial work (due to the threading subtleties that you already know)  that I prefer to take on once we move to NET 4.6 (otherwise the fix will have to be postponed, and it's pretty important for troubleshooting)